### PR TITLE
feat: EVM support with Frontier

### PR DESCRIPTION
**Pull request by vgantchev**
This PR implements support for EVM using the [Frontier pallets by Moonbeam](https://github.com/moonbeam-foundation/frontier)

- [x] Implement Frontier into hydradx-runtime
- [x] Implement Frontier support into RPC
- [x] Add tests
- [x] account conversion
- [x] Runtime configuration
- [x] final check, versions and merge

## Gas
`FeeCalculator` uses `FixedGasPrice` with `DEFAULT_BASE_FEE_PER_GAS = 0.5 gwei`

Not using [pallet-base-fee](https://github.com/moonbeam-foundation/frontier/blob/master/frame/base-fee/src/lib.rs) because it does not support minimum gas price in the currently used version `polkadot-v0.9.38`

After upgrading the Runtime to `polkadot-v1.1.0`, `FeeCalculator` should be switched to [pallet-base-fee](https://github.com/moonbeam-foundation/frontier/blob/master/frame/base-fee/src/lib.rs)

## Precompiles
A precompile is native Substrate code that has an Ethereum-style address and can be called using the Ethereum API. The precompiles allow us to call the Substrate runtime directly.

There are precompiles added to hydraDX runtime:
- dispatch precompile to execute any extrinsic
- multicurrency precompile to implement the ERC20 standard, to be used in MetaMask

### CODE REVIEW

File: node/src/evm.rs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+use cumulus_client_consensus_common::ParachainBlockImportMarker;
+use fc_consensus::FrontierBlockImport;
+use fc_db::Backend as FrontierBackend;
+use fc_mapping_sync::{MappingSyncWorker, SyncStrategy};
+use fc_rpc::{EthTask, OverrideHandle};
+use fc_rpc_core::types::{FeeHistoryCache, FilterPool};
+use fp_rpc::EthereumRuntimeRPCApi;
+use fp_storage::EthereumStorageSchema;
+use futures::{future, StreamExt};
+use hydradx_runtime::Block;
```
Comment (mrq1911) [inline node/src/evm.rs:40]: primitives Blocks should be used instead probably

File: integration-tests/Cargo.toml [hunk 2]
```diff
@@ -57,6 +57,10 @@ orml-xtokens = { workspace = true }
 orml-xcm-support = { workspace = true }
 orml-unknown-tokens = { workspace = true }
 
+# Evm
+pallet-evm = { workspace = true }
+evm-runtime = { git = "https://github.com/rust-blockchain/evm", rev = "13240a8a551586fdef0b5028ed73af80b248092a", default-features = false }
+fp-evm = { git = "https://github.com/justinphamnz/frontier", branch = "polkadot-v0.9.38", default-features = false }
```
Comment (jak-pan) [inline integration-tests/Cargo.toml:63]: Do we use this?

File: integration-tests/Cargo.toml [hunk 3]
```diff
@@ -57,6 +57,10 @@ orml-xtokens = { workspace = true }
 orml-xcm-support = { workspace = true }
 orml-unknown-tokens = { workspace = true }
 
+# Evm
+pallet-evm = { workspace = true }
+evm-runtime = { git = "https://github.com/rust-blockchain/evm", rev = "13240a8a551586fdef0b5028ed73af80b248092a", default-features = false }
```
Comment (jak-pan) [inline integration-tests/Cargo.toml:62]: Why are we using this implementation?

File: Cargo.toml [hunk 4]
```diff
@@ -200,6 +207,27 @@ cumulus-test-relay-sproof-builder = { git = "https://github.com/galacticcouncil/
...
+fp-self-contained = { git = "https://github.com/PureStake/frontier", branch = "moonbeam-polkadot-v0.9.38", default-features = false }
+fp-storage = { git = "https://github.com/PureStake/frontier", branch = "moonbeam-polkadot-v0.9.38", default-features = false }
+
+pallet-ethereum = { git = "https://github.com/PureStake/frontier", branch = "moonbeam-polkadot-v0.9.38", default-features = false }
+pallet-evm = { git = "https://github.com/PureStake/frontier", branch = "moonbeam-polkadot-v0.9.38", default-features = false }
+pallet-evm-chain-id = { git = "https://github.com/PureStake/frontier", branch = "moonbeam-polkadot-v0.9.38", default-features = false }
+pallet-evm-precompile-dispatch = { git = "https://github.com/PureStake/frontier", branch = "moonbeam-polkadot-v0.9.38", default-features = false }
+
+#Evm from acala
+module-evm-utility-macro = { path = "runtime/hydradx/src/evm/modules/evm-utility/macro", default-features = false}
```
[the remark this answers is not in the document]
Comment (dmoka) [reply Cargo.toml:231]: it is used to generate function selector for enum values, for example in multicurrency precompiles: https://github.com/galacticcouncil/HydraDX-node/blob/21a3672765aec9f9c56e9d8c838fc4614e2704d2/runtime/hydradx/src/evm/precompiles/multicurrency.rs#L43

So we can read the selector from MM input then based on the selector we can do the action:

https://github.com/galacticcouncil/HydraDX-node/blob/21a3672765aec9f9c56e9d8c838fc4614e2704d2/runtime/hydradx/src/evm/precompiles/multicurrency.rs#L85

File: Cargo.toml [hunk 5]
```diff
@@ -35,9 +35,14 @@ members = [
   'math',
   'pallets/staking',
   'pallets/democracy',
+  'runtime/hydradx/src/evm/modules/evm-utility/macro',
+  #'runtime/hydradx/src/evm/modules/evm',
```
[the remark this answers is not in the document]
Comment (dmoka) [reply Cargo.toml]: not needed, i remove it

File: runtime/hydradx/src/evm/precompiles/costs.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+//         .!Y7Y#@@#:   ?@@@G?JJ^           Built with <3 for decentralisation.
+//            !G@@@Y    .&@@&J:
+//              ^5@#.   7@#?.               Copyright (C) 2021-2023  Intergalactic, Limited (GIB).
+//                :5P^.?G7.                 SPDX-License-Identifier: Apache-2.0
+//                  :?Y!                    Licensed under the Apache License, Version 2.0 (the "License");
+//                                          you may not use this file except in compliance with the License.
+//                                          http://www.apache.org/licenses/LICENSE-2.0
+
+//! Cost calculations.
+//! TODO: PR EVM to make those cost calculations public.
```
Comment (vgantchev) [inline runtime/hydradx/src/evm/precompiles/costs.rs:23]: @dmoka there is a todo here, still relevant?

File: runtime/hydradx/src/evm/precompiles/costs.rs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+use crate::evm::precompiles::EvmResult;
+use pallet_evm::{ExitError, PrecompileFailure};
+
+// TODO CHANGE
```
Comment (vgantchev) [inline runtime/hydradx/src/evm/precompiles/costs.rs:28]: and here @dmoka

File: runtime/hydradx/src/evm/precompiles/multicurrency.rs [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+	<Runtime as pallet_asset_registry::Config>::AssetId: core::convert::From<AssetId>,
+	Currencies: MultiCurrency<Runtime::AccountId, CurrencyId = AssetId, Balance = Balance>,
+	pallet_currencies::Pallet<Runtime>: MultiCurrency<Runtime::AccountId, CurrencyId = AssetId, Balance = Balance>,
+	<Runtime as frame_system::Config>::AccountId: core::convert::From<sp_runtime::AccountId32>,
+	<<Runtime as frame_system::Config>::RuntimeCall as Dispatchable>::RuntimeOrigin: OriginTrait,
+{
+	fn execute(handle: &mut impl PrecompileHandle) -> pallet_evm::PrecompileResult {
+		let address = handle.code_address();
+		if let Some(asset_id) = HydraErc20Mapping::decode_evm_address(address) {
+			log::debug!(target: "evm", "multicurrency: currency id: {:?}", asset_id);
```
Comment (vgantchev) [inline runtime/hydradx/src/evm/precompiles/multicurrency.rs:73]: @dmoka shall we cleanup the debugger?

File: traits/src/registry.rs [hunk 9]
```diff
@@ -18,6 +19,14 @@ pub trait Registry<AssetId, AssetName, Balance, Error> {
...
 }
 
+pub trait RegistryQueryForEvm<AssetId, NameStringLimit, Error> {
+	fn retrieve_asset_name(name: AssetId) -> Result<BoundedVec<u8, NameStringLimit>, Error>;
+
+	fn retrieve_asset_symbol(name: AssetId) -> Result<BoundedVec<u8, NameStringLimit>, Error>;
+
+	fn retrieve_asset_decimals(name: AssetId) -> Result<u8, Error>;
+}
+
```
Comment (enthusiastmartin) [inline traits/src/registry.rs:29]: we already have traits that returns decimals - it is called InspectRegistry.. We should not really create a new trait to do the same. 

we can just extend that trait to return name and symbol too

btw.  I saw you use some pallets directly in evm impl,  so you could use pallet_registry directly as well, without the need of any traits.

Comment (vgantchev) [reply traits/src/registry.rs]: Resolved by https://github.com/galacticcouncil/HydraDX-node/pull/646/commits/9215cb78aded2bf00a51cc72d5297c4c99c115dd and following commits

File: runtime/hydradx/src/evm/modules/evm-utility/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,32 @@
+[package]
```
Comment (mrq1911) [inline runtime/hydradx/src/evm/modules/evm-utility/Cargo.toml:1]: @dmoka why we just don't depend on acala module?

File: runtime/hydradx/src/evm/modules/evm-utility/macro/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,80 @@
+//                    :                     $$\   $$\                 $$\                    $$$$$$$\  $$\   $$\
```
Comment (mrq1911) [inline runtime/hydradx/src/evm/modules/evm-utility/macro/src/lib.rs:1]: this shouldn't be in the copied acala files probably

Comment (mrq1911) [reply runtime/hydradx/src/evm/modules/evm-utility/macro/src/lib.rs:1]: you have to acknowledge where you took the code

File: runtime/hydradx/src/evm/precompiles/multicurrency.rs [hunk 12]
```diff
@@ -0,0 +1,250 @@
+//                    :                     $$\   $$\                 $$\                    $$$$$$$\  $$\   $$\
```
Comment (mrq1911) [inline runtime/hydradx/src/evm/precompiles/multicurrency.rs:1]: this file is copied from acala as well?

**Review by mrq1911 — CHANGES_REQUESTED** [review]

File: runtime/hydradx/src/evm/precompiles/multicurrency.rs (hunk 12 above)
Comment (dmoka) [reply runtime/hydradx/src/evm/precompiles/multicurrency.rs:1]: it is not copied, and i think the header is fine for this. Or what is the issue with it?

Comment (mrq1911) [reply runtime/hydradx/src/evm/precompiles/multicurrency.rs:1]: nice

**Review by mrq1911 — APPROVED** [review]

**Pull request merged**